packages/shared/src/components/onboarding/OnboardingFeedHeader.tsx

```diff
@@ -0,0 +1,161 @@
+  ButtonIconPosition,
+  ButtonSize,
+  ButtonVariant,
+} from '../buttons/Button';
+import { REQUIRED_TAGS_THRESHOLD } from './common';
+import { ArrowIcon } from '../icons';
+import { LayoutHeader } from '../layout/common';
+import { usePrompt } from '../../hooks/usePrompt';
+import { useMyFeed } from '../../hooks/useMyFeed';
+import { useAlertsContext } from '../../contexts/AlertContext';
+import { useAnalyticsContext } from '../../contexts/AnalyticsContext';
+import { AnalyticsEvent } from '../../lib/analytics';
+
+type OnboardingFeedHeaderProps = {
+  isPreviewFeedVisible: boolean;
+  setPreviewFeedVisible: React.Dispatch<React.SetStateAction<boolean>>;
+  isFeedPreviewEnabled: boolean;
+  tagsCount: number;
+};
+
+export const OnboardingFeedHeader = ({
+  isPreviewFeedVisible,
+  setPreviewFeedVisible,
+  isFeedPreviewEnabled,
+  tagsCount,
+}: OnboardingFeedHeaderProps): JSX.Element => {
+  let stopNav = true;
+  const { updateAlerts } = useAlertsContext();
+  const { trackEvent } = useAnalyticsContext();
+  const { showPrompt } = usePrompt();
+  const router = useRouter();
+  const { registerLocalFilters } = useMyFeed();
+
+  const completeOnboarding = useCallback(() => {
+    registerLocalFilters();
+    updateAlerts({ filter: false, myFeed: 'created' });
+
+    trackEvent({
+      event_name: AnalyticsEvent.CreateFeed,
+    });
+
+    return router.replace({
+      pathname: router.route,
+      query: {
+        welcome: 'true',
+      },
+    });
+  }, [registerLocalFilters, router, trackEvent, updateAlerts]);
+
+  const preventNavigation = useCallback(
+    (newRoute?: string) => {
+      showPrompt({
+        title: 'Discard tag selection?',
+        description: 'Your personalized feed is only a few tags away',
+        cancelButton: {
+          title: 'Stay',
+          variant: ButtonVariant.Primary,
+          color: ButtonColor.Cabbage,
+        },
+        okButton: {
+          title: 'Leave',
+          variant: ButtonVariant.Secondary,
+        },
+        className: {
+          buttons: 'flex-row-reverse',
+        },
+      }).then((leave) => {
+        if (leave) {
+          // eslint-disable-next-line react-hooks/exhaustive-deps
+          stopNav = false;
+          trackEvent({
+            event_name: AnalyticsEvent.OnboardingSkip,
+          });
+          router.replace(newRoute);
+        }
+      });
+    },
+    [router, showPrompt],
+  );
+
+  const routeHandler = useCallback(
+    (newRoute: string) => {
+      if (!stopNav || newRoute === '/my-feed') {
+        return;
+      }
+      router.events.emit('routeChangeError');
+      preventNavigation(newRoute);
+      throw new Error('Cancelling navigation');
+    },
+    [preventNavigation, router.events, stopNav],
+  );
+
+  useEffect(() => {
+    router.events.on('routeChangeStart', routeHandler);
+    return () => {
+      router.events.off('routeChangeStart', routeHandler);
+    };
+  }, [preventNavigation, routeHandler, router.events, stopNav]);
```

Comment:
Better to move `routeHandler` inside `useEffect` since they care about same dependencies. `preventNavigation` can also go inline then.

Reply:
Also like this the `routeHandler` reference can change before useEffect cleanup which would leak event handlers.

Reply:
Refactored it now, and still works as intended 🙏